### DISCUSSION TITLE
Enable PSR-12 compliance in addons

### DISCUSSION THREAD

**Discussion by acalvino4** [ExpressionEngine/ExpressionEngine#3105] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
The plugin setup [documentation](https://docs.expressionengine.com/latest/development/plugins.html#creating-the-plugin-file) says that plugin names should be snake_case, and that the `pi.` file and the class therein must be named the same. However, Expression Engine also [aims](https://expressionengine.com/blog/expressionengine-adopts-psr-12) to be PSR-12 compliant, which requires class names to be CapitalCamelCase. If we really mean to aim for PSR-12 compliance, we should shoot for it to be possible within the framework. (Note: currently it is technically possible, if you ignore the plugin documentation and name your addon `MyAddon` and refer to it in your templates as `{exp:MyAddon}`)

**Describe the solution you'd like**
We should either update the documentation to recommend CapitalCamelCase for addon names, or else change the code which registers plugins to map the snake_case name to CapitalCamelCase when looking for the `pi.` file path and the class.

**Describe alternatives you've considered**
The alternative is to ignore one of the two pieces of documentation I linked at the top.

**Teachability, Documentation, Adoption, Migration Strategy**
If the selected solution is a mapping of snake_case to CapitalCamelCase, while allowing the current behavior to continue working as well, this would look like
* your plugin folder is named `my_addon`
* plugin references in your template are `{exp:my_addon}`
* your plugin namespaces is `MyPrefix\MyAddon`
* your plugin file is named `pi.MyAddon.php` (or `pi.my_addon.php` - I'm somewhat indifferent, though my preference would be for the former since it comes closer to a 1-1 classname-filename correspondence)
* your plugin class is `MyAddon`
* similar behavior for other types of addons

I am also fine with the solution of updating the documentation to recommend CapitalCamelCase plugin names. That would probably be the easiest solution.

- [x] I am capable and would like to work on implementation of this feature if it is considered. I have already identified most parts of the code that would need changing.

**Comment by intoeetive** [comment]
Unfortunately the add-ons architecture has been developed long time before PSR-12 even existed, so we can't really change it without breaking existing add-ons.

If you are using VSCode, you can use `phpcs.ruleset.xml` ruleset in the root of EE repo, which adds few exceptions to the rule, so that the add-on class names do not generate warning

**Comment by acalvino4** [comment]
We don't necessarily need to change the architecture. CaptitalCamelCase already works for addon names, hence one of my proposed solutions was just 
> update the documentation to recommend CapitalCamelCase for addon names

The only potentially undesired consequence of this is that referring to an addon in your template looks like `{exp:MyAddonName}` instead of `{exp:my_addon_name}`, but I don't think that's a very significant issue. It doesn't break compatibility because this only goes for addons that choose to follow PSR-12; not all plugins are aiming to. In other words, it's up to addon developers to follow PSR-12 or not, but I don't see any reason that the documentation shouldn't recommend it as the default.

If we really did want to maintain the use of snake_case for addons in a template, it shouldn't be too hard for those references to first look for a folder in addons of the same name, and for the CapitalCamelCase version only if that fails.
